README.md

```diff
@@ -1 +1,101 @@
-# edge-block-service
\ No newline at end of file
+# minibus 🚐
+
+> A edge block service to store IPFS data structures
+
+## Getting started
+
+One time set up of your cloudflare worker subdomain for dev:
+
+- `npm install` - Install the project dependencies from the monorepo root directory
+- Sign up to Cloudflare and log in with your default browser.
+- `npm i @cloudflare/wrangler -g` - Install the Cloudflare wrangler CLI
+- `wrangler login` - Authenticate your wrangler cli; it'll open your browser.
+- Copy your cloudflare account id from `wrangler whoami`
+- Update `wrangler.toml` with a new `env`. Set your env name to be the value of `whoami` on your system you can use `pnpm start` to run the worker in dev mode for you.
+
+  [**wrangler.toml**](./wrangler.toml)
+
+  ```toml
+  [env.bobbytables]
+  workers_dev = true
+  account_id = "<what does the `wrangler whoami` say>"
+  ```
+
+- Add secrets
+
+  ```sh
+    wrangler secret put SENTRY_DSN --env $(whoami) # Get from Sentry (not required for dev)
+    wrangler secret put LOGTAIL_TOKEN --env $(whoami) # Get from Logtail
+    wrangler secret put SALT --env $(whoami) # open `https://csprng.xyz/v1/api` in the browser and use the value of `Data`
+  ```
+
+- Add R2 bucket (Note that it is only available as Private Beta at the time of writing)
+
+  ```sh
+  wrangler r2 bucket create blockstore --env $(whoami)
+  # 🌀  Creating bucket "blockstore"
+  # ✨  Success!
+  ```
+
+- `npm run publish` - Publish the worker under your env. An alias for `wrangler publish --env $(whoami)`
+- `npm dev` - Run the worker in dev mode. An alias for `wrangler dev --env $(whoami)`
+
+You only need to `npm dev` for subsequent runs. PR your env config to the `wrangler.toml` to celebrate 🎉
+
+## High level architecture
+
+The Edge HTTP internal API to store IPFS data structures on R2.
+
+![High level Architecture](./minibus.web3.storage-api.jpg)
+
+## Usage
+
+You can see full HTTP API specification at https://web3.storage/block-service-docs.
+
+### 🔒 `PUT /`
+
+> Store individuals blocks by multihash in R2.
+
+TODO: curl example
+
+### 🔒 `GET /:multihash`
+
+> Multihash block read interface.
+
+The block data is retrieved and discovered by multihash, in order to enable the same data to be referred to with different codecs.
+
+```sh
+curl -X GET -H 'Authorization: Bearer ACCESS_KEY' https://minibus.web3.storage/bciqjhirzogurjzpkzpykrusrktg2gcodyhds7o4zctkhyyhtznublca
+```
+
+### 🔒 `PUT /car`
```

Comment:
I think `PUT /cid` and `GET /cid` might be a better interface, I realize @mikeal's rational for storing / indexing by multihash, but that does not imply that should be read / write interface necessarily. Here is what I'd suggest:

1. Make REST interface be `GET /cid` and `PUT /cid`.
2. Underlying storage should key by `multihash`, basically removing codec / cid version info from `cid`
3. `PUT /cid` will also able to verify that provided data matches the `multihash` of the CID and fail bytes don't match e.g. if something in the middle scrambled them.
4. For specific codecs we could add additional features like with car CID. In `PUT` is for CAR CID it can do what comment suggest below.
5. `GET /car_cid` will end up returning bytes for the car.

This seems more intuitive and we could plug other codec specific smarts in the future.

Reply:
Added the redirect from received CID to multihash as previously discussed.

I would suggest creating and issue and iterate on follow up PR for the bit shifts part - receive param with either multihash or CID.
I agree that we should do that, but that opens other details to be taken into account. Like, if a cid in PUT has different hashing algorithm we need to validate it, but still compute multihash to R2 key using sha256 to not duplicate content stored.